File: src/auth/services/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { UsersService } from '../../users/users.service';
import * as bcrypt from "bcrypt"
import { PayLoadToken } from '../interface/auth.interface';
import { User } from '@prisma/client';
import { Request, Response } from 'express';
import { AuthServiceProps } from '../interface/authService.interface';

@Injectable()
export class AuthService implements AuthServiceProps {
    constructor(
        private readonly jwtService: JwtService,
        private readonly userService: UsersService,
    ) { }

    public async generateToken(user: User, res: Response): Promise<any> {
        const payload: PayLoadToken = { roles: user.roles, sub: user.id }

        const expirationDate = 24 * 60 * 60;
        const RefreshExpirationDate = 7 * 24 * 60 * 60;

        const access_token = this.jwtService.sign(payload, {
            secret: process.env.SECRET_KEY,
            expiresIn: `${expirationDate}`
        })
        const refresh_token = this.jwtService.sign(payload, {
            secret: process.env.REFRESH_TOKEN_KEY,
            expiresIn: `${RefreshExpirationDate}`
        })
        const hashRefresh_token = await bcrypt.hash(refresh_token, 10)
        await this.userService.updateToken(user.id, hashRefresh_token)

        res.cookie("Authentication", access_token, {
            httpOnly: true,
            secure: process.env.NODE_ENV === "production",
            expires: new Date(Date.now() + expirationDate * 1000)
        })
        res.cookie("Refresh", refresh_token, {
            httpOnly: true,
            secure: process.env.NODE_ENV === "production",
            expires: new Date(Date.now() + RefreshExpirationDate * 1000)
        })
        return { access_token, user }
    }

    public async validationUser(username: string, password: string): Promise<User> {
        const user = await this.userService.finBy({ key: "username", value: username })

        const match = await bcrypt.compare(password, user.password)
        if (!match) throw new UnauthorizedException("Password is Wrong")

        return user
    }

    public async veryfyRefreshToken(refreshToken: string, userId: number): Promise<User> {
        const user = await this.userService.findOne(userId)
        if (!user) throw new UnauthorizedException("User not found")

        const authenticated = await bcrypt.compare(refreshToken, user.refreshToken)
        if (!authenticated) throw new UnauthorizedException()

        return user
    }

    public async profile(req: Request): Promise<{ username: string }> {
        const token = req.cookies["Authentication"]
        if (!token) throw new UnauthorizedException("No token found in cookies")

        const decodeToken = this.jwtService.verify(token, { secret: process.env.SECRET_KEY })
        const userid = decodeToken.sub

        const user = await this.userService.findOne(userid)
        return ({ username: user.username })
    }
    
    public async logOut(userId: number, res: Response): Promise<any> {
        await this.userService.updateToken(userId, null)

        res.cookie("Authenticated", "", {
            expires: new Date(0),
            httpOnly: true,
            secure: process.env.NODE_ENV === "production"
        })
        res.cookie("Refresh", "", {
            expires: new Date(0),
            httpOnly: true,
            secure: process.env.NODE_ENV === "production"
        })
    }
}

File: TASK-FLOW.NET/Auth/Service/AuthService.cs
using TASK_FLOW.NET.Auth.DTO;
using TASK_FLOW.NET.Auth.Service.Interface;
using TASK_FLOW.NET.User.Model;
using TASK_FLOW.NET.User.Service.Interface;
using BCrypt.Net;
using Microsoft.AspNetCore.Identity;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using TASK_FLOW.NET.Auth.JWT.Service;
using TASK_FLOW.NET.Auth.Cookie.Service;
using TASK_FLOW.NET.Auth.JWT.Service.Interface;
using TASK_FLOW.NET.Auth.Cookie.Service.Interface;
namespace TASK_FLOW.NET.Auth.Service
{
    public class AuthService : IAuthService
    {
        private readonly string secretKey;
        private readonly IUserService _userService;
        private readonly IHttpContextAccessor _httpContext;
        private readonly ITokenService _tokenService;
        private readonly ICookieService _cookieService;

        public AuthService(IConfiguration config, IUserService userService, IHttpContextAccessor httpContext, ITokenService tokenService, ICookieService cookieService)
        {
            this.secretKey = config.GetSection("JwtSettings").GetSection("secretKey").ToString();
            this._userService = userService;
            this._httpContext = httpContext;
            this._tokenService = tokenService;
            this._cookieService = cookieService;
        }

        public async Task<string> GenerateToken(UsersModel body)
        {
            var token = this._tokenService.GenerateToken(body);
            await this._userService.UpdateToken(body.Id, token.RefreshTokenHasher);

            this._cookieService.SetTokenCookies(this._httpContext.HttpContext.Response, token);
            return token.AccessToken;
            //var keyBytes = Encoding.UTF8.GetBytes(secretKey);
            //var tokenHandler = new JwtSecurityTokenHandler();

            //var claims = new List<Claim>
            //{
            //    new Claim ("sub", body.Id.ToString()),
            //    new Claim ("rol", body.Roles.ToString())
            //};

            //var tokenDescriptor = new SecurityTokenDescriptor
            //{
            //    Subject = new ClaimsIdentity(claims),
            //    Expires = DateTime.UtcNow.AddDays(2),
            //    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(keyBytes), SecurityAlgorithms.HmacSha256Signature)
            //};
            //var refreshTokenDescriptor = new SecurityTokenDescriptor
            //{
            //    Subject = new ClaimsIdentity(claims),
            //    Expires = DateTime.UtcNow.AddDays(5),
            //    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(keyBytes), SecurityAlgorithms.HmacSha256Signature)
            //};

            //var tokenConfig = tokenHandler.CreateToken(tokenDescriptor);
            //var refreshTokenConfig = tokenHandler.CreateToken(refreshTokenDescriptor);

            //var accessToken = tokenHandler.WriteToken(tokenConfig);
            //var refreshToken = tokenHandler.WriteToken(refreshTokenConfig);

            //var hashRefreshToken = BCrypt.Net.BCrypt.HashPassword(refreshToken, 10);

            //await this._userService.UpdateToken(body.Id, hashRefreshToken);

            //this._httpContext.HttpContext.Response.Cookies.Append("Authentication", accessToken, new CookieOptions
            //{
            //    HttpOnly = true,
            //    Secure = true,
            //    Expires = DateTime.UtcNow.AddDays(2),
            //    SameSite = SameSiteMode.Strict
            //});

            //this._httpContext.HttpContext.Response.Cookies.Append("RefreshToken", refreshToken, new CookieOptions
            //{
            //    HttpOnly = true,
            //    Secure = true,
            //    Expires = DateTime.UtcNow.AddDays(5),
            //    SameSite = SameSiteMode.Strict
            //});

            //return accessToken;
        }

        public async Task<string> GetProfile()
        {
            var user = await this.GetUserByCookie();
            return user.Username;
        }

        public async Task<UsersModel> GetUserByCookie()
        {
            //var token = this._httpContext.HttpContext.Request.Cookies["Authentication"];
            //if (token == null) throw new UnauthorizedAccessException("Token not found");

            //var tokenHandler = new JwtSecurityTokenHandler();
            //var jwtToken = tokenHandler.ReadToken(token) as JwtSecurityToken;

            //var userId = int.Parse(jwtToken?.Claims.FirstOrDefault(c => c.Type == "sub")?.Value);

            //if (userId == null) throw new UnauthorizedAccessException("Invalid Token");
            var token = this._cookieService.GetCookies(this._httpContext.HttpContext.Request, "Authentication");
            if (string.IsNullOrEmpty(token)) throw new UnauthorizedAccessException("token not found");
            var userId = this._tokenService.ValidateToken(token, "sub");

            var user = await this._userService.GetById(userId);
            return user;
        }

        public async Task Logout()
        {
            var user = await this.GetUserByCookie();
            await this._userService.UpdateToken(user.Id, null);

            this._cookieService.ClearTokenCookies(this._httpContext.HttpContext.Response);
            //this._httpContext.HttpContext.Response.Cookies.Append("Authentication", "", new CookieOptions
            //{
            //    HttpOnly = true,
            //    Secure = true,
            //    Expires = DateTime.UnixEpoch,
            //    SameSite = SameSiteMode.Strict
            //});
            //this._httpContext.HttpContext.Response.Cookies.Append("RefreshToken", "", new CookieOptions
            //{
            //    HttpOnly = true,
            //    Secure = true,
            //    Expires = DateTime.UnixEpoch,
            //    SameSite = SameSiteMode.Strict
            //});
        }

        public async Task<UsersModel> RefreshTokenValidate(string refreshToken, int id)
        {
            var user = await this._userService.GetById(id);

            var match = BCrypt.Net.BCrypt.Equals(refreshToken, user.RefreshToken);
            if (match == null) throw new UnauthorizedAccessException();

            return user;
        }

        public async Task<UsersModel> ValidationUser(AuthPropsDTO body)
        {
            var user = await this._userService.FindByAuth("Username", body.Username);
            var passwordHasher = new PasswordHasher<UsersModel>();

            var verificationResult = passwordHasher.VerifyHashedPassword(user, user.Password, body.Password);

            if (verificationResult == PasswordVerificationResult.Failed) throw new UnauthorizedAccessException("Password wrong");

            return user;
        }
    }
}
